File: leetcode/78-subsets/subsets.cpp
class Solution {
public:
    set<vector<int>> st;
    vector<vector<int>> subsets(vector<int>& nums) {
         solve(nums,0,{});
         vector<vector<int>> ans;
         for(auto it : st){
            ans.push_back(it);
         }
        return ans;
    }

    void solve(vector<int>& nums,int i,vector<int> curr){
        if(i == nums.size()){
            st.insert(curr);
            return;
        }

        // two option either current element will be set either it will be not

        solve(nums,i+1,curr);
        curr.push_back(nums[i]);
        solve(nums,i+1,curr);
        curr.pop_back();

        return;
    }
};

File: leetcode/78-subsets/subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> results = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        solve(nums,0,results,path);

        return results;
        
    }

    private void solve(int[] nums,int i,List<List<Integer>> results,List<Integer> path){
           if(i == nums.length){
                results.add(new ArrayList<>(path));
                return;
           }

           solve(nums,i+1,results,path);
           path.add(nums[i]);
           solve(nums,i+1,results,path);
           path.remove(path.size()-1);
    }
}